src/solvers/sundials.pyx

```diff
@@ -1898,10 +1898,16 @@ cdef class CVode(Explicit_ODE):
                 raise CVodeError(flag, self.t)
             
     def initialize_event_detection(self):
-        def event_func(t, y):
-            return self.problem.state_events(t, y, self.sw)
-        self.event_func = event_func
-        self.g_old = self.event_func(self.t, self.y)
+        if self.options["external_event_detection"]:
+            def event_func(t, y):
+                return 0, self.problem.state_events(t, y, self.sw)
+            self.event_func = event_func
+            _, self.g_old = self.event_func(self.t, self.y)
+        else:
+            def event_func(t, y):
+                return self.problem.state_events(t, y, self.sw)
+            self.event_func = event_func
+            self.g_old = self.event_func(self.t, self.y)
```

Comment:
1. Is there any performance gain/loss depending on if event_func is a cython function or a regular python function? I would assume not but curious how it works.
2. Can we add any comment/docstring to the two functions? It is not clear to me why the first one returns `0, ...` while the other one is just returning the state events.

Reply:
1. There may be a minor performance gain. It would appear that a _def_ function translates the underlying C code and a wrapper function that corresponds to the _def_ signature, a _cdef_ function would skip the wrapper part. (https://stackoverflow.com/questions/28362009/definition-of-def-cdef-and-cpdef-in-cython#:~:text=cdef%20functions%20are%20quicker%20to,to%20call%20it%20from%20Python).) Might lead to a minor performance improvement. Would require some re-structuring though, since cdef function cannot be defined inside other functions.
2. Added comments, please check if this is sufficient. Note, this came with [https://github.com/modelon-community/Assimulo/pull/53](https://github.com/modelon-community/Assimulo/pull/53 ). This can be improved in CVode.

Reply:
Looks great, and thank you for the reply!